client/src/components/reader/Section/ResourceViewer/Wrapper.js

```diff
@@ -0,0 +1,94 @@
+import React, { PureComponent, PropTypes } from 'react';
+import Viewer from './Viewer';
+import throttle from 'lodash/throttle';
+
+export default class ResourceViewerWrapper extends PureComponent {
+
+  static displayName = "ResourceViewer.Wrapper";
+
+  static propTypes = {
+    resources: PropTypes.array,
+    containerSize: PropTypes.number,
+    currentResources: PropTypes.array,
+    body: PropTypes.object
+  };
+
+  constructor() {
+    super();
+    this.state = {
+      availableResourceMarkers: [],
+      viewableResourceMarkers: []
+    };
+
+    this.updateAvailableResourceMarkers = this.updateAvailableResourceMarkers.bind(this);
+    this.updateViewableResourceMarkers = this.updateViewableResourceMarkers.bind(this);
+
+    this.throttledUpdateViewableResources = throttle(() => {
+      this.updateViewableResourceMarkers();
+    }, 200);
+  }
+
+  componentDidMount() {
+    // Run once on mount ?
+    this.updateAvailableResourceMarkers();
+    this.updateViewableResourceMarkers();
+
+    window.addEventListener('scroll', this.throttledUpdateViewableResources);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('scroll', this.throttledUpdateViewableResources);
+  }
+
+  componentDidUpdate(prevProps) {
+    // Update available resource markers if we have new props or state
+    if (prevProps !== this.props) {
+      this.updateAvailableResourceMarkers();
+      this.updateViewableResourceMarkers();
+    }
+  }
+
+  updateAvailableResourceMarkers() {
+    if (this.props.body) {
+      // resourceMarkers.forEach((marker)=> {
+      //   const resourceIds = marker.getAttribute('data-resources');
+      //   availableMarkers = availableMarkers.concat(resourceIds.split(','));
+      // });
+      const markerNodes = this.props.body.querySelectorAll('[data-resources]');
+      this.setState({
+        availableResourceMarkers: [...markerNodes]
+      });
+    }
+  }
+
+  updateViewableResourceMarkers() {
+    const markers = this.state.availableResourceMarkers;
+    const viewableMarkers = [];
+    markers.forEach((marker) => {
+      const rect = marker.getBoundingClientRect();
+      if (rect.top > 0 && rect.top < window.innerHeight) {
+        const resourceIds = marker.getAttribute('data-resources').split(',');
+        resourceIds.forEach((id) => {
+          viewableMarkers.push({
+            id,
+            top: rect.top
+          });
+        });
+      }
+    });
+
+    this.setState({
+      viewableResourceMarkers: viewableMarkers
```

Comment:
If you were to call this variable `viewableResourceMarkers` instead of `viewableMarkers` then this could be more concise as well: `this.setState({ viewableResourceMarkers });`